Restructuring of fparser2 PR from issue #92

rupertford opened the pull request.

rupertford commented:
I'm taking this PR back for a bit as I've noticed a few formatting errors and missing coverage.

rupertford commented:
OK, ready for 1st review @arporter. I've had problems with a couple of the tests which I've had to make xfail. If you have any suggestions that would be great!

arporter commented on src/fparser/two/tests/test_parser.py:
```diff
@@ -0,0 +1,71 @@
+
+
+@pytest.mark.xfail(reason="This works in isolation but fails with other "
+                   "tests. No idea why :-(")
```
Wow, you're hitting lots of exciting problems in this Issue. I bet this is something to do with the fact that pytest probably imports all the modules it finds before running the tests. In other modules you do, e.g.:

    # this is required to setup the fortran2003 classes
    _ = ParserFactory().create(std="f2003")

Possibly this should be done as part of test set-up (c.f. a fixture?) rather than every time a module is imported? The fact that I can run e.g. `py.test test_comments.py test_parser.py` without trouble (presumably because test_comments sets-up the 2003 parser) is evidence for this.

rupertford replied:
There were two problems. The first was that the setting up of class relationships in the subclass dictionary was not being reset before each ParserFactory().create() call which meant that the list of classes just got longer and longer each time we called create(). This caused an f2003 parser to accept fortran2008 after an f2008 parser had been instantiated. This is now fixed. The second problem is what the author suspects, but this only came to light in one f2008 test when I changed the last parser test to end with the creation of an f2003 parser. I've now changed all f2008 tests to  call ParserFactory().create(std="f2008"). Actually in practice only one test needed this to be added to make all failing tests to work but the fact it was this one looks random to me so changing all f2008 tests seemed the best option.

arporter commented on src/fparser/two/tests/test_parser.py:
```diff
@@ -0,0 +1,71 @@
+    code = str(ast)
+    assert "SUBMODULE (x) y\nEND SUBMODULE y" in code
+
+    with pytest.raises(SystemExit):
```
As mentioned earlier, this should be an Exception of some sort, not a system exit.

rupertford replied:
Please see previous answer.

arporter replied:
OK. Please add a TODO here.

rupertford commented on doc/developers_guide.rst:
```diff
@@ -80,3 +49,225 @@ provided as a string. Both of these classes sub-class `FortranReaderBase`:
+In fparser2 each rule is implemented in a class with the class names
+closely following the rule names. For example, `program` is
+implemented by the `Program` class and `program-unit` is implemented
+by the `Program_Unit` class. In general if '-' is replaced with '_'
```
Modified. Thanks.

rupertford commented on doc/developers_guide.rst:
```diff
@@ -80,3 +49,225 @@ provided as a string. Both of these classes sub-class `FortranReaderBase`:
+and each word is capitalised then that is the name of the equivalent
+class.
+
+The Fortran2003, classes exist in the Fortran2003.py file and the
```
Removed.

rupertford commented on doc/developers_guide.rst:
```diff
@@ -80,3 +49,225 @@ provided as a string. Both of these classes sub-class `FortranReaderBase`:
+
+As the above method is a simple choice between different classes, the
+appropriate static match method is already implemented in one of the
+base classes (`Base`) and therefore is not required. Therefore to
```
Modified, with a slight change (overridden->written)

rupertford commented on doc/developers_guide.rst:
```diff
@@ -80,3 +49,225 @@ provided as a string. Both of these classes sub-class `FortranReaderBase`:
+++++++++++++++++
+
+Some classes that are specified as strings in the `subclass_names` or
+`use_names` variables are not required to be implemented as
```
Much better. Done.

rupertford commented on doc/developers_guide.rst:
```diff
@@ -80,3 +49,225 @@ provided as a string. Both of these classes sub-class `FortranReaderBase`:
+2) classes of the form '\*\_List'
+3) classes of the form 'Scalar\_\*'
+
+The reason for this is that such classes can be written in a generic
```
Done thanks.

rupertford commented on src/fparser/two/Fortran2008.py:
```diff
@@ -0,0 +1,437 @@
+# Modified work Copyright (c) 2018 Science and Technology
```
The problem is that I have copied some of Pearu's original code to the end of the file that creates classes so whilst it is a new file, some of the code has been copied from an existing file. I think therefore that I need to keep the old copyright notice.

arporter replied:
Ah yes, thanks for explaining. There must be some way of avoiding this duplication but I think that can be left for another Issue.

rupertford commented on src/fparser/two/parser.py:
```diff
@@ -0,0 +1,262 @@
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+# DAMAGE.
+
+'''This file provides utilities to create a fortran parser suitable
```
A little picky but a fair point. Done.

rupertford commented on src/fparser/two/parser.py:
```diff
@@ -0,0 +1,262 @@
+
+    def create(self, std=None):
+        '''Creates a class hierarchy suitable for the specified fortran
+        standard
```
Done.

rupertford commented on src/fparser/two/parser.py:
```diff
@@ -0,0 +1,262 @@
+        >>> from fparser.two.parser import ParserFactory
+        >>> f2003_parser = ParserFactory().create()
+        >>> f2003_parser = ParserFactory().create(std='f2003')
+        >>> f2008_parser = ParserFactory().create(std='f2008')
```
I am not sure about this. I think that in general a class docstring should just show how to create an instance of the class. However, I've relented and added something.

rupertford commented on src/fparser/two/parser.py:
```diff
@@ -0,0 +1,262 @@
+            # we return the Fortran2003 one.
+            return Fortran2003.Program
+        else:
+            print ("unsupported standard {0}".format(std))
```
Exceptions need to be sorted in fparser2 in general and we have an issue to look at this #52. Therefore I decided to simply exit for the time being, rather than spending time creating a new exception that we would need to change. Is that a good enough argument or would you like me to add an exception in any case?

arporter replied:
That's a good enough argument but please add a TODO linking to #52.

rupertford commented on src/fparser/two/parser.py:
```diff
@@ -0,0 +1,262 @@
+        class_type = type(fparser.two.Fortran2003.Base)
+
+        for clsinfo in input_classes:
+            clsname = "{0}.{1}".format(clsinfo[1].__module__, clsinfo[0])
```
Agreed. However, I have purposely left this code as-is for the moment as this pr is primarily about restructuring fparser2 to support adding in f2008. I plan to have a separate issue to tidy up and document the code that generates the class relationships. If the reviewer is happy with this I will create a new issue and add these comments to it.

arporter replied:
Yes, that's a good plan.

rupertford commented on src/fparser/two/parser.py:
```diff
@@ -0,0 +1,262 @@
+
+        if 1:  # Optimize subclass tree:
+
+            def _rpl_list(clsname):
```
See previous comment.

arporter replied:
OK...so long as our current understanding is stored somewhere so it doesn't fall down the cracks between Issues :-)

rupertford commented on src/fparser/two/parser.py:
```diff
@@ -0,0 +1,262 @@
+                    error_string = 'Not implemented: {0}'.format(clsname)
+                    logging.getLogger(__name__).debug(error_string)
+                    return []
+                # remove this code when all classes are implemented
```
See previous comment.

rupertford commented on src/fparser/two/parser.py:
```diff
@@ -0,0 +1,262 @@
+                bits = fparser.two.Fortran2003.Base.subclasses[clsname]
+            except KeyError:
+                fparser.two.Fortran2003.Base.subclasses[clsname] = bits = []
+            for name in subclass_names:
```
See previous comment.

rupertford commented on src/fparser/two/parser.py:
```diff
@@ -0,0 +1,262 @@
+                              format(name, clsname)
+                    logging.getLogger(__name__).debug(message)
+
+        if 1:
```
See previous comment.

rupertford commented on src/fparser/two/tests/fortran2008/test_end_submodule_stmt_r1119.py:
```diff
@@ -0,0 +1,101 @@
+
+# All rights reserved.
+
+# Modifications made as part of the fparser project are distributed
```
I think it already does have the new license text.

arporter replied:
Oh yes. I must have been thrown by the "Modifications made as part of..." bit.

rupertford commented on src/fparser/two/tests/fortran2008/test_end_submodule_stmt_r1119.py:
```diff
@@ -0,0 +1,101 @@
+_ = ParserFactory().create(std="f2008")
+
+
+def test_simple_1():
```
Done.

rupertford commented on src/fparser/two/tests/fortran2008/test_end_submodule_stmt_r1119.py:
```diff
@@ -0,0 +1,101 @@
+    assert "End_Submodule_Stmt: 'end submod'" in str(excinfo.value)
+
+
+@pytest.mark.xfail(reason="Does not raise an error if there is extra "
```
Actually the f2003.py code works fine in this situation. I had made a mistake in the assert text so the assert was failing - doh!

arporter replied:
Oh! I wonder how that works?

rupertford commented on src/fparser/two/tests/fortran2008/test_parent_identifier_r1118.py:
```diff
@@ -0,0 +1,121 @@
+
+
+@pytest.mark.xfail(reason="Does not raise an error when there are too "
+                   "many colons")
```
Brill. Test no longer fails :-)

rupertford commented on src/fparser/two/tests/fortran2008/test_submodule_r1116.py:
```diff
@@ -0,0 +1,248 @@
+
+
+def test_submodule_samename():
+    '''Test the parsing of a submodule with the same name : C1114'''
```
Yes, that is clearer thanks.

rupertford commented on src/fparser/two/tests/fortran2008/test_submodule_stmt_r1117.py:
```diff
@@ -0,0 +1,129 @@
+    assert "Submodule_Stmt: 'submodule (id) name'" in str(excinfo.value)
+
+
+@pytest.mark.xfail(reason="Does not raise an error if there is extra "
```
Again it was my mistake with the contents of the matching string in the assert.

rupertford commented:
I've hopefully addressed the reviewers comments. In two cases I've recommended creating new issues rather than making the suggested change(s). The first is adding an exception and the second is restructuring existing code. Back to @arporter.

arporter commented on src/fparser/two/tests/fortran2008/test_end_submodule_stmt_r1119.py:
```diff
@@ -41,21 +41,21 @@
 def test_simple_1():
-    '''Test the parsing of a minimal end-submodule statement'''
+    '''Test the parsing of a minimal end-submodule statement.'''
+    _ = ParserFactory().create(std="f2008")
```
It would be better to use a set-up fixture in this test file so that you only have one place where you create the f2008 parser.

rupertford replied:
Done. Created a fixture in a conftest.py file and use it for all tests that previously explicitly used ParserFactory.

arporter commented on src/fparser/two/tests/fortran2008/test_parent_identifier_r1118.py:
```diff
@@ -41,81 +41,82 @@
 def test_simple_1():
-    '''Test the parsing of a minimal parent-identifier statement'''
+    '''Test the parsing of a minimal parent-identifier statement.'''
+    _ = ParserFactory().create(std="f2008")
```
Again, please could you do this just once in a fixture in this file.

rupertford replied:
Done.

arporter commented on src/fparser/two/tests/fortran2008/test_program_unit_r202.py:
```diff
@@ -42,16 +42,15 @@
+    program-unit options.
 
     '''
+    _ = ParserFactory().create(std="f2008")
```
Please do this just once in a fixture.

rupertford replied:
Done.

arporter commented on src/fparser/two/tests/fortran2008/test_submodule_r1116.py:
```diff
@@ -52,12 +52,11 @@
 def test_submodule():
-    '''Test the parsing of a minimal submodule'''
+    '''Test the parsing of a minimal submodule.'''
+    _ = ParserFactory().create(std="f2008")
```
Fixture please.

rupertford replied:
Done.

arporter commented on src/fparser/two/tests/fortran2008/test_submodule_r1116.py:
```diff
@@ -238,11 +249,14 @@ def test_submodule_samename():
 def test_submodule_differentname():
-    '''Test and exception is raised if a submodule has a different name :
-    C1114'''
+    '''Test and exception is raised if the end submodule statement has a
```
s/and/an/

rupertford replied:
Well spotted. Done.

arporter commented on src/fparser/two/tests/fortran2008/test_submodule_stmt_r1117.py:
```diff
@@ -41,89 +41,98 @@
 
 def test_simple():
     ''' Test the parsing of a submodule statement'''
+    _ = ParserFactory().create(std="f2008")
```
Fixture please.

rupertford replied:
Done.

arporter commented on src/fparser/two/tests/fortran2008/test_submodule_stmt_r1117.py:
```diff
@@ -41,89 +41,98 @@
+
     def mockreturn(my_string):
-        return ["XXX", "", ""]
+        ''' Mock function for the splitparen function.'''
```
Rather than define `mockreturn()`, you could just use a lambda for this:

    lambda x: ["XXX", "", ""]

but it's not a big deal.

rupertford replied:
Not a big deal but the code is simpler so I've changed it. Thanks.

rupertford commented:
I think I've addressed all the reviewers comments. I've also created two new issues for code restructuring that was deferred in this PR. Ready for 3rd review. Back to @arporter.
